src/aws-cpp-sdk-core/include/smithy/client/AwsSmithyClientBase.h

```diff
@@ -148,6 +148,8 @@ namespace client
                                       std::shared_ptr<Aws::Http::HttpResponse> httpResponse) const;
 
         inline virtual const char* GetServiceClientName() const { return m_serviceName.c_str(); }
+        inline virtual std::shared_ptr<Aws::Http::HttpClient> GetHttpClient() { return m_httpClient; }
```

Comment:
return const ref
returning object will increase the shared_ptr use ref and the code actually using it won't be able to determine if this http client got only a single owner

Reply:
I did this because this is how we currently handle it in [AwsClient](https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/include/aws/core/client/AWSClient.h#L172) and its how you added [GetHttpClient](https://github.com/aws/aws-sdk-cpp/commit/9e8e71d3f2b263e6194a00a08c39f227b7d56970#diff-579705b19505b3ce9d73df235130af7e4455c41396dd18f00956c64cfa51beb8) when you made the PR for improving shutdown. I had to add this method when doing the rebase. Im not opposed to making this change but we will have to update [AWSClientAsyncCRTP](https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/include/aws/core/client/AWSClientAsyncCRTP.h#L90-L124) where we call shutdown to create a new shutdown method that could be a template specialization for what `GetHttpClient` returns whether its a pointer or ref.

but curious if we should make it const ref also for AwsClient so we dont have to diverge the codepath for smithy client vs aws client shutdown.

Reply:
I meant to add const ref as in AwsClient you linked
```
const std::shared_ptr<Aws::Http::HttpClient>&
```
vs
```
std::shared_ptr<Aws::Http::HttpClient>
```
Returning shared_ptr by value will increase it is internal reference count.
I had to perform such check in CRTP
```
            if (pClient->GetHttpClient().use_count() == 1)
            {
                pClient->DisableRequestProcessing();
            }
```
to be sure that the service client is not (accidentally) shallow-copied and the shutdown would not accidentally block some other client.
Unfortunately we did not enforce any policy on (default implicit shallow) copy-constructors, so have to rely on use_count.

Reply:
ahhhhhhh, thank you, makes sense